integration: Add tests for port mappings

akerouanton opened the pull request:
**- What I did**

Introduce new integration tests that ensure port mappings are correctly accessible from the host, and from a remote host. This is another attempt at upstreaming some of the tests introduced in:

- https://github.com/moby/moby/pull/45850

These tests should help ensure nothing breaks when we'll modify iptables rules, or when we'll add nftables support.

**- How I did it**

I added a new `L3Segment` struct that simulates (using netns) a LAN segment where arbitrary `Host`s live. That's what makes `TestAccessPublishedPortFromRemoteHost` possible.

That specific test simulates two hosts: one is the 'current' netns where the daemon-under-test and the tests themselves are executed. The other is the 'remote host'. Each host gets a pair of veth. A third network namespace is used to isolate the bridge from any network programming the engine does.

To create all these objects (netns, veth, etc...), simple iproute2 invocations are made. The previous attempt was using netlink calls, which was an horrendously bad idea since there was no way to recreate the same env for manual debugging. All iproute2 invocations are logged, so now it's just a matter of copy / pasting logs into a shell script, tidying them up a bit, and then running the script.

The `L3Segment` struct has one important method: `AddHost`.

And `Host` has two important methods:

- `Run`, which runs a command in the host's netns. That command is logged too.
- `Do`, which runs a callback in the host's netns.

Both `Host`s and `L3Segment` can take any number of IP addrs, making it easy to test v4/v6-only, as well as, dual-stack scenarios in one go.

This should be enough to cover all use-cases.

**- How to verify it**

CI is green.

akerouanton committed:
inte/networking: move port mapping tests into a dedicated file

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
inte/networking: test accessing a published port from the host

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

robmry commented on integration/networking/port_mapping_linux_test.go:
```diff
@@ -0,0 +1,313 @@
+				container.WithNetworkMode(bridgeName))
+			defer c.ContainerRemove(ctx, serverID, containertypes.RemoveOptions{Force: true})
+
+			poll.WaitOn(t, container.IsInState(ctx, c, serverID, "running"), poll.WithDelay(100*time.Millisecond))
```
I think the container's set to running before the POST to start it returns? (If it's needed here, the WaitOn is probably needed in a lot of other tests.) If the container goes straight to exited or some other state, without an error from the start request, I think this will just wait forever?

robmry commented on integration/networking/port_mapping_linux_test.go:
```diff
@@ -0,0 +1,313 @@
+			serverID := container.Run(ctx, t, c,
+				container.WithName(sanitizeCtrName(t.Name()+"-server")),
+				container.WithExposedPorts("80/tcp"),
+				container.WithPortMap(nat.PortMap{"80/tcp": {{HostPort: serverPort}}}),
```
As the mapping doesn't include a `HostIP`, the container's port 80 should be accessible via any host address (apart from `--userland-proxy=false --ipv6=false`, which will be IPv4 only).

If that's the intention, perhaps it'd be worth checking access via each of the host addresses in each test... dropping the last component of the test-config so that, for each combo of userland-proxy and IPv6-enabled, there's an http request to each of eth0 and lo's v4 and v6 addresses.

But it might also be worth testing that, with a host address in the port mapping, it's not possible to access the http server via other host addresses?

robmry commented on integration/networking/port_mapping_linux_test.go:
```diff
@@ -0,0 +1,313 @@
+		t.Run(tc.name, func(t *testing.T) {
+			ctx := testutil.StartSpan(ctx, t)
+
+			clientAddr, err := getIfaceAddress(t, tc.clientIface, tc.ipv6)
```
`clientAddr` -> `hostAddr` ... as it's a server address, on the host? (Ditto for `clientIface` -> `hostIface`.)

akerouanton committed:
inte/networking: test access to published ports from remote host

Along with this new test, a new struct `L3Segment` is introduced to
simulate hosts connected on a same switched network. This struct will
let us test various scenarios where published ports and containers
should or should not be accessible from remote hosts.

The new test introduced, `TestAccessPublishedPortFromRemoteHost`, skips
link-local address as port publishing doesn't work on those addresses
currently. This will be fixed in a future commit.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

thaJeztah commented on integration/networking/port_mapping_linux_test.go:
```diff
@@ -178,3 +204,96 @@ func TestProxy4To6(t *testing.T) {
+
+	ctx := setupTest(t)
+
+	assert.NilError(t, exec.Command("sysctl", "net.ipv6.conf.all.disable_ipv6=0").Run())
+	out, err := exec.Command("ip", "addr", "add", "fdfb:5cbb:29bf::2/64", "dev", "eth0", "nodad").CombinedOutput()
```
Is this something that needs to be reverted after the test? Can this cause issues when running these tests on the host (not in a container)?

neersighted reviewed (approved):
LGTM overall with the exception of @thaJeztah's concern about sysctls, which he beat me to.

thaJeztah reviewed (approved):
LGTM
